models.py

```diff
@@ -0,0 +1,385 @@
+#!/usr/bin/env python
+
+from collections import OrderedDict
+import numpy as np
+from scipy import ndimage
+import torch  
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.autograd import Variable
+import torchvision
+import matplotlib.pyplot as plt
+import time
+
+
+class reactive_net(nn.Module):
+
+    def __init__(self, use_cuda): # , snapshot=None
+        super(reactive_net, self).__init__()
+        self.use_cuda = use_cuda
+
+        # Initialize network trunks with DenseNet pre-trained on ImageNet
+        self.push_color_trunk = torchvision.models.densenet.densenet121(pretrained=True)
+        self.push_depth_trunk = torchvision.models.densenet.densenet121(pretrained=True)
+        self.grasp_color_trunk = torchvision.models.densenet.densenet121(pretrained=True)
+        self.grasp_depth_trunk = torchvision.models.densenet.densenet121(pretrained=True)
+
+        self.num_rotations = 16
+
+        # Construct network branches for pushing and grasping
+        self.pushnet = nn.Sequential(OrderedDict([
+            ('push-norm0', nn.BatchNorm2d(2048)),
+            ('push-relu0', nn.ReLU(inplace=True)),
+            ('push-conv0', nn.Conv2d(2048, 64, kernel_size=1, stride=1, bias=False)),
+            ('push-norm1', nn.BatchNorm2d(64)),
+            ('push-relu1', nn.ReLU(inplace=True)),
+            ('push-conv1', nn.Conv2d(64, 3, kernel_size=1, stride=1, bias=False))
+            # ('push-upsample2', nn.Upsample(scale_factor=4, mode='bilinear'))
+        ]))
+        self.graspnet = nn.Sequential(OrderedDict([
+            ('grasp-norm0', nn.BatchNorm2d(2048)),
+            ('grasp-relu0', nn.ReLU(inplace=True)),
+            ('grasp-conv0', nn.Conv2d(2048, 64, kernel_size=1, stride=1, bias=False)),
+            ('grasp-norm1', nn.BatchNorm2d(64)),
+            ('grasp-relu1', nn.ReLU(inplace=True)),
+            ('grasp-conv1', nn.Conv2d(64, 3, kernel_size=1, stride=1, bias=False))
+            # ('grasp-upsample2', nn.Upsample(scale_factor=4, mode='bilinear'))
+        ]))
+
+        # Initialize network weights
+        for m in self.named_modules():
+            if 'push-' in m[0] or 'grasp-' in m[0]:
+                if isinstance(m[1], nn.Conv2d):
+                    nn.init.kaiming_normal(m[1].weight.data)
+                elif isinstance(m[1], nn.BatchNorm2d):
+                    m[1].weight.data.fill_(1)
+                    m[1].bias.data.zero_()
+
+        # Initialize output variable (for backprop)
+        self.interm_feat = []
+        self.output_prob = []
+
+
+    def forward(self, input_color_data, input_depth_data, is_volatile=False, specific_rotation=-1):
```

Comment:
What's the difference between is_volatile being True or False in this function besides the fact that the variables if is_volatile is True are local while if it's False, the variables are a part of the class?